packages/same-structure/src/match.js

```diff
@@ -0,0 +1,196 @@
+ *
+ * @param {Passable} passable
+ */
+const patternKindOf = passable => {
+  const passStyle = passStyleOf(passable);
+  if (passStyle === 'copyRecord' && PATTERN_KIND in passable) {
+    const patternKind = passable[PATTERN_KIND];
+    assert.string(patternKind);
+    return patternKind;
+  }
+  return undefined;
+};
+harden(patternKindOf);
+
+/**
+ * A *passable* object is a pass-by-copy superstructure ending in
+ * non-pass-by-copy leaves, each of which is either a promise or a
+ * REMOTE_STYLE. A passable object in which none of the leaves are promises
+ * is a *comparable*. A comparable object in which none of the
+ * copyRecords are pattern records is *ground*. All other comparables are
+ * *non-ground*.
+ *
+ * Only some contexts care about the distinction between ground and non-ground,
+ * such as the arguments to the `match` function below. For most other purposes,
+ * these are simply passable and comparable objects. However, some uses of
+ * `sameStructure` to compare comparables should probably either be guarded by
+ * `isGround` tests or converted to `match` calls.
+ *
+ * @param {Passable} passable
+ * @returns {boolean}
+ */
+function isGround(passable) {
+  const passStyle = passStyleOf(passable);
+  switch (passStyle) {
+    case 'null':
+    case 'undefined':
+    case 'string':
+    case 'boolean':
+    case 'number':
+    case 'bigint':
+    case REMOTE_STYLE:
+    case 'copyError': {
+      return true;
+    }
+    case 'promise': {
+      return false;
+    }
+    case 'copyArray': {
+      return passable.every(isGround);
+    }
+    case 'copyRecord': {
+      const patternKind = patternKindOf(passable);
+      if (patternKind !== undefined) {
+        return false;
+      }
+      return values(passable).every(isGround);
+    }
+    default: {
+      throw new TypeError(`unrecognized passStyle ${passStyle}`);
+    }
+  }
+}
+harden(isGround);
+
+/**
+ * @param {Pattern} outerPattern
+ * @param {Ground} outerSpecimen
+ * @returns {Bindings | undefined}
+ */
+function match(outerPattern, outerSpecimen) {
+  assert(
+    isGround(outerSpecimen),
+    d`Can only match against ground comparables for now`,
+  );
+
+  // Although it violates Jessie, don't harden `bindings` yet
+  const bindings = { __proto__: null };
+
+  // TODO Reduce redundancy with sameStructure.
+  function matchInternal(pattern, specimen) {
+    const patternKind = patternKindOf(pattern);
+    if (patternKind !== undefined) {
+      switch (patternKind) {
+        case '*': {
+          // wildcard. matches anything.
+          return true;
+        }
+        case 'bind': {
+          const { name } = pattern;
+          // binds specimen to bindings[name]
+          assert.string(name);
+          if (name in bindings) {
+            // Note: sameStructure rather than match, as both sides came from
+            // the outerSpecimen.
+            return sameStructure(bindings[name], specimen);
+          }
+          bindings[name] = specimen;
```

Comment:
`name` is used as a computed property name. For all the same safety reasons, we should assert that it is a keyword.

Reply:
Since this is just exploratory for now, I just added a TODO to that effect. Resolving this comment in favor of that TODO.